iotilecore/iotile/core/hw/debug/firmware_image_analyzer.py

```diff
@@ -0,0 +1,155 @@
+        """Returns firmware module version specified in cdb app info
+        """
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return (self.cdb_app_info['module_major_version'],
+                self.cdb_app_info['module_minor_version'],
+                self.cdb_app_info['module_patch_version'])
+
+    def get_cdb_app_info(self):
+        """Gets the string of bytes of the CDB app info
+        """
+        cdb_app_info_bytes = self._hex_image.gets(self.max_addr - (FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO - 1), \
+                FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO)
+        if self._check_cdb_app_info(cdb_app_info_bytes):
+            self.cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+            return self.cdb_app_info
+
+        addresses = self._hex_image.addresses()
+        for address_index in range(0, self.num_addresses - 4):
+            search_bytes = self._hex_image.gets(addresses[address_index], 4)
+            search, = struct.unpack("<L", search_bytes)
+            if search == self.magic_number:
+                cdb_app_info_bytes = self._hex_image.gets(addresses[address_index - FirmwareImageAnalyzer.MAGIC_NUMBER_BYTE_OFFSET], \
+                    FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO)
+                if self._check_cdb_app_info(cdb_app_info_bytes):
+                    self.cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+                    return self.cdb_app_info
+                
+
+    def _check_cdb_app_info(self, cdb_app_info_bytes):
+        """Helper function to check the byte string if it is a valid CDB app info
+        """
+        _hardware_type, _api_major_version, _api_minor_version, name, _module_major_version, \
+            _module_minor_version, _module_patch_version, _num_slave_commands, _num_required_configs, \
+            _num_total_configs, _reserved, _p_config_variables, _p_slave_handlers, magic_number, \
+            _firmware_checksum = struct.unpack("<BBB6sBBBBBBBLLLL", cdb_app_info_bytes)
+        cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+
+        try:
+            name = cdb_app_info['name'].decode("utf-8")
+        except:
+            return False
+
+        if name in FirmwareImageAnalyzer.KNOWN_FIRMWARE_NAMES \
+            and cdb_app_info['magic_number'] == self.magic_number:
+            return True
+
+        return False
+
+    def _format_cdb_app_info(self, cdb_app_info_bytes):
+        """Helper function to format the cdb_app_info byte string to dict
+        """
+        hardware_type, api_major_version, api_minor_version, name, module_major_version, \
+            module_minor_version, module_patch_version, num_slave_commands, num_required_configs, \
+            num_total_configs, reserved, p_config_variables, p_slave_handlers, magic_number, \
+            firmware_checksum = struct.unpack("<BBB6sBBBBBBBLLLL", cdb_app_info_bytes)
+
+        cdb_app_info = dict()
```

Comment:
This is a valid way to populate the dictionary, but you can also initialize it in place with something like

```python
cdb_app_info = { 'hardware_type': hardware_type,
                             'api_major_version': api_major_version,
```
                            ...
                           }

Reply:
Fixed